### PR TITLE
Fix query analyzer logic for mixed conjunctions of terms and ranges

### LINKED ISSUE #49684: [CI] org.elasticsearch.percolator.CandidateQueryTests.testDuel2 failure
This test seems to have been failing recently:

master: https://gradle-enterprise.elastic.co/s/eel3j36akwpam
7.x:  https://gradle-enterprise.elastic.co/s/xjdre63zy2hmk

It does not reproduce locally. Failure:

```

org.elasticsearch.percolator.CandidateQueryTests > testDuel2 FAILED
--
java.lang.AssertionError:
Expected: <6L>
but: was <5L>
at __randomizedtesting.SeedInfo.seed([D22432087DDAD7A4:E884A4C0AB3467C]:0)
at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
at org.junit.Assert.assertThat(Assert.java:956)
at org.junit.Assert.assertThat(Assert.java:923)
at org.elasticsearch.percolator.CandidateQueryTests.duelRun(CandidateQueryTests.java:1062)
at org.elasticsearch.percolator.CandidateQueryTests.testDuel2(CandidateQueryTests.java:376


```

### PR DESCRIPTION
**Pull request by romseygeek**
When the query analyzer examines a conjunction containing both terms and ranges,
it should only include ranges in the `minimum_should_match` calculation if there are no
other range queries on that same field within the conjunction.  This is because we cannot
build a selection query over disjoint ranges on the same field, and it is not easy to check
if two range queries have an overlap.

The current logic to calculate this just sets `minimum_should_match` to `1` or `0`, dependent
on whether or not the current range is over a field that has already been seen.  However, this
can be incorrect in the case that there are terms in the same match group which adjust the
`minimum_should_match` downwards.  Instead, the logic should be changed to match the
terms extraction, whereby we adjust `minimum_should_match` downwards if we have already
seen a range field.

Fixes #49684

### CODE REVIEW

File: modules/percolator/src/test/java/org/elasticsearch/percolator/QueryAnalyzerTests.java [hunk 1]
```diff
@@ -1208,4 +1208,51 @@ public void testIntervalQueries() {
...
+
+        Query disj = new BooleanQuery.Builder()
+            .add(IntPoint.newRangeQuery("i", 0, 10), Occur.SHOULD)
+            .add(new TermQuery(new Term("f", "v1")), Occur.SHOULD)
+            .add(new TermQuery(new Term("f", "v1")), Occur.SHOULD)
+            .setMinimumNumberShouldMatch(2)
+            .build();
+
+        Result r = analyze(disj, Version.CURRENT);
+        assertThat(r.minimumShouldMatch, equalTo(1));
```
Comment (jpountz) [inline modules/percolator/src/test/java/org/elasticsearch/percolator/QueryAnalyzerTests.java]: can you count the number of extractions and check `verified` for all queries?

File: modules/percolator/src/test/java/org/elasticsearch/percolator/QueryAnalyzerTests.java [hunk 2]
```diff
@@ -1208,4 +1208,51 @@ public void testIntervalQueries() {
...
+        Query q2 = new BooleanQuery.Builder()
+            .add(new TermQuery(new Term("f", "v1")), Occur.FILTER)
+            .add(IntPoint.newRangeQuery("i", 15, 20), Occur.SHOULD)
+            .add(new TermQuery(new Term("f", "v2")), Occur.SHOULD)
+            .add(new TermQuery(new Term("f", "v2")), Occur.MUST)
+            .setMinimumNumberShouldMatch(1)
+            .build();
+
+        result = analyze(q2, Version.CURRENT);
+        assertThat(result.minimumShouldMatch, equalTo(2));
```
Comment (jpountz) [inline modules/percolator/src/test/java/org/elasticsearch/percolator/QueryAnalyzerTests.java]: let's maybe also have a test for the case that there are multiple range queries on the same field, or multiple range queries on different fields?

**Review by jpountz — COMMENTED** [review]
This logic is tricky. :)

**Comment by romseygeek** [comment]
Thanks for the review @jpountz .  I added more tests, and discovered that the logic was *still* not quite correct - tricky, as you say...  We now only check that range fields from a particular set of extractions have not been seen in other results.  Multiple identical range fields within a single Result extraction set will have already been dealt with, so don't require any adjustment to the msm.

File: modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java [hunk 3]
```diff
@@ -286,6 +284,9 @@ private static Result handleConjunction(List<Result> conjunctionsWithUnknowns) {
                     verified = false;
                 }
             }
+            // add range fields from this Result to the seenRangeFields set so that minimumShouldMatch is correctly
+            // calculated for subsequent Results
+            result.extractions.stream().filter(e -> e.range != null).forEach(e -> seenRangeFields.add(e.range.fieldName));
```
Comment (jpountz) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java]: nit: I generally have a preference for using method refs when possible and using new lines when chaining, e.g.
```suggestion
            result.extractions.stream()
                .filter(Objects::nonNull)
                .map(e -> e.range)
                .forEach(seenRangeFields::add);
```

**Review by jpountz — APPROVED** [review]
LGTM. It's hard for me to reason about all possible corner cases, but the tests give me some confidence that it's at least more correct than the previous logic.

**Pull request merged**